### PR TITLE
(Database) Serial scanning for Gamecube/MegaCD/SegaCD/Saturn/PSX/PSP/Dreamcast/Wii

### PR DESCRIPTION
**Pull request by pkos**
## Guidelines

1. Rebase before opening a pull request
2. If you are sending several unrelated fixes or features, use a branch and a separate pull request for each
3. If possible try squashing everything in a single commit. This is particularly beneficial in the case of feature merges since it allows easy bisecting when a problem arises

## Description

These updates improve support for "Directory Scan".  While previously the scan was made against raw disc serials, this code converts and parses those serials into redump.org curated serials.

Logs and playlists now include these Redump serials.

Multi-disc support.

CRC checks are still enabled and disabling them may increase performance.

Bad discs are scrubbed and logged.

## Related Issues

@RobLoach has updated the libretro-databases to include serials (3/17/20).
https://github.com/libretro/libretro-database/pull/1026

@RobLoach has updated some string parsing to redump codes.  This added support for Playstation - Portable as confirmed no changes to the existing psp detect serial function was needed.
https://github.com/RobLoach/libretro-dats/pull/37

@RobLoach has updated muti-disc appendices (3/28/20).  This update added support for multi-disc scanning.
https://github.com/libretro/libretro-database/pull/1032/

This PR code will close the following issues:
https://github.com/libretro/RetroArch/issues/10282
https://github.com/libretro/libretro-database/issues/990
https://github.com/libretro/libretro-database/issues/839
https://github.com/libretro/libretro-database/issues/813
https://github.com/libretro/libretro-database/issues/613
https://github.com/libretro/libretro-database/issues/423

Currently, scanner will not detect .gcz, .cso files, but it will crc them.
https://github.com/libretro/RetroArch/issues/7428

This PR code replaces changes already merged here:
Reworked magic numbers.
https://github.com/libretro/RetroArch/issues/6544
https://github.com/libretro/RetroArch/pull/6555
https://github.com/libretro/RetroArch/issues/5239
Serials are in playlists.
https://github.com/libretro/RetroArch/issues/6548

## Related Pull Requests

The following libchdr issue will crash the scanner.
https://github.com/libretro/RetroArch/issues/10368
Issue resolved (3/31/20)
https://github.com/libretro/RetroArch/pull/10376

## Reviewers

@RealNC
@Celerizer
@Alcaro
@panasonic
@phcoder 

They have added support in #programming, fixing bugs and providing C code.

### CODE REVIEW

File: tasks/task_database_cue.c [hunk 1]
```diff
@@ -19,6 +19,12 @@
 #include <ctype.h>
 #include <string.h>
 
+// regex
+#include <iostream>
+#include <string>
+#include <regex>
+#include <iterator>
```
Comment (RobLoach) [inline tasks/task_database_cue.c]: Don't think these includes are needed.

**Comment by RobLoach** [comment]
Excited to see scanning improve here. Nicely done! Is there something in redump dats that needs to change?

**Comment by pkos** [comment]
> Excited to see scanning improve here. Nicely done! Is there something in redump dats that needs to change?

Thank you, I just updated this PR with truly real c code... yay.

**Comment by pkos** [comment]
> > Excited to see scanning improve here. Nicely done! Is there something in redump dats that needs to change?
> 
> Thank you, I just updated this PR with truly real c code... yay.

The redump dats are fine.  The game_id variable (serial) is set in this code to match the redump dats serials.  If anything, maybe an update to the gamecube rdb would be nice.

**Comment by pkos** [comment]
> > > Excited to see scanning improve here. Nicely done! Is there something in redump dats that needs to change?
> > 
> > 
> > Thank you, I just updated this PR with truly real c code... yay.
> 
> The redump dats are fine. The game_id variable (serial) is set in this code to match the redump dats serials. If anything, maybe an update to the gamecube rdb would be nice.

@RealNC has helped me to properly use C code formatting.. i think its ready

**Comment by realnc** [comment]
You still have errors. I assumed you actually compiled the core prior to pushing it:

`return rue;`

File: tasks/task_database_cue.c [hunk 2]
```diff
@@ -303,17 +304,78 @@ int detect_psp_game(intfstream_t *fd, char *game_id)
...
 
-   return rv;
+   /** convert raw gamecube serial to redump serial.
+   not enough is known about the disc data to properly
+   convert every raw serial to redump serial.  it will
+   only fail with the following excpetions:  the disc
+   is a multi disc game, and the subregions of europe
+   P-UKV, P-AUS, X-UKV, X-EUU will not match redump.**/
+   
+   // insert prefix
```
Comment (RobLoach) [inline tasks/task_database_cue.c]:
``` suggestion
   /* insert prefix */
```

File: tasks/task_database_cue.c [hunk 3]
```diff
@@ -303,17 +304,78 @@ int detect_psp_game(intfstream_t *fd, char *game_id)
...
+   int x = 7;
+   strncpy(pre_game_id, prefix, x);
+   pre_game_id[x] = '\0';
+   strcat(pre_game_id, raw_game_id);
+
+   // NYI: This perl code checks filename looking for multi disc to add to the serial
```
Comment (RobLoach) [inline tasks/task_database_cue.c]:
``` suggestion
   /* NYI: This perl code checks filename looking for multi disc to add to the serial */
```

File: tasks/task_database_cue.c [hunk 4]
```diff
@@ -303,17 +304,78 @@ int detect_psp_game(intfstream_t *fd, char *game_id)
...
+   /** if ($element =~ m/\(Disc 1/i or $element =~ m/\(Disk 1/i) {
+      $game_id = $game_id . "-0";
+   } elsif ($element =~ m/\(Disc 2/i or $element =~ m/\(Disk 2/i) {
+      $game_id = $game_id . "-1";
+   } **/
+
+   // check region then insert region suffix
```
Comment (RobLoach) [inline tasks/task_database_cue.c]:
``` suggestion
   /* check region then insert region suffix */
```

**Comment by pkos** [comment]
> You still have errors. I assumed you actually compiled the code prior to pushing it:
> 
> `return rue;`
> 
> You should compile and test your code before submitting.

fixed and just recently compiled on windows

File: retroarch.cfg [hunk 5]
```diff
@@ -1,915 +0,0 @@
-## Skeleton config file for RetroArch
```
Comment (RobLoach) [inline retroarch.cfg]: Strange... `retroarch.cfg` is in [.gitignore#L20](https://github.com/libretro/RetroArch/blob/master/.gitignore#L20), but this file is still committed into the repo. While I do agree we should either remove it, or make sure it's not in .gitignore, I feel like this change likely shouldn't happen here.

File: tasks/task_database_cue.c (hunk 1 above)
Comment (twinaphex) [reply tasks/task_database_cue.c]: Indeed, please don't use any C++ headers in C code.

Comment (pkos) [reply tasks/task_database_cue.c]: I fixed that, and no more c++ code, just C, it compiles with  make C89_BUILD=1.  I've tested the scanner with Robs new rdb files and its working.

**Comment by pkos** [comment]
> You still have errors. I assumed you actually compiled the code prior to pushing it:
> 
> `return rue;`
> 
> You should compile and test your code before submitting.

Definately, the code compiles with  make C89_BUILD=1.. more than that its actually works with the latest rdbs from Rob.

File: Makefile.psp1 [hunk 6]
```diff
@@ -25,7 +25,7 @@ INCDIR = deps deps/stb deps/7zip deps/pthreads deps/pthreads/platform/psp deps/p
 CFLAGS = $(OPTIMIZE_LV) -G0 -std=gnu99 -ffast-math -fsingle-precision-constant
 ASFLAGS = $(CFLAGS)
 
-RARCH_DEFINES = -DPSP -D_MIPS_ARCH_ALLEGREX -DHAVE_LANGEXTRA -DHAVE_ZLIB -DHAVE_RPNG -DHAVE_RJPEG -DHAVE_GRIFFIN=1 -DRARCH_INTERNAL -DRARCH_CONSOLE -DHAVE_MENU -DHAVE_CONFIGFILE -DHAVE_RGUI -DHAVE_FILTERS_BUILTIN -DHAVE_7ZIP -DHAVE_CC_RESAMPLER
+RARCH_DEFINES = -DPSP -D_MIPS_ARCH_ALLEGREX -DHAVE_LANGEXTRA -DHAVE_ZLIB -DHAVE_AUDIOMIXER -DHAVE_RPNG -DHAVE_RJPEG -DHAVE_GRIFFIN=1 -DRARCH_INTERNAL -DRARCH_CONSOLE -DHAVE_MENU -DHAVE_CONFIGFILE -DHAVE_RGUI -DHAVE_FILTERS_BUILTIN -DHAVE_7ZIP -DHAVE_CC_RESAMPLER
```
Comment (RobLoach) [inline Makefile.psp1]: Looks like some unrelated changes. Did the rebase mess up again?

Comment (pkos) [reply Makefile.psp1]: yes, but i restored the original files

**Comment by pkos** [comment]
> This pull request **introduces 3 alerts** when merging [f58fcb2](https://github.com/libretro/RetroArch/commit/f58fcb23852e44def0d04ef6c1ee893859ea71ca) into [f48c668](https://github.com/libretro/RetroArch/commit/f48c6684477c2c926e47ed6ae991c2ae74d180a4) - [view on LGTM.com](https://lgtm.com/projects/g/libretro/RetroArch/rev/pr-a631b827121f6a7f009ad51e1864f1aa435e468b)
> 
> **new alerts:**
> 
> * 3 for Comparison result is always the same

its not always the same, based on raw serial!!!!!!

**Comment by pkos** [comment]
> This pull request **introduces 3 alerts** when merging [bee2073](https://github.com/libretro/RetroArch/commit/bee2073f2a0dee61a5973dab0cb4ca2107235166) into [9729b5a](https://github.com/libretro/RetroArch/commit/9729b5a917aa72f4cfe28b979e1b1aa6ad081254) - [view on LGTM.com](https://lgtm.com/projects/g/libretro/RetroArch/rev/pr-3d7c156e475addd90f9fb9db5087e731f39b5f25)
> 
> **new alerts:**
> 
> * 3 for Comparison result is always the same

Fixed

**Comment by RobLoach** [comment]
Cleaned up a few things over at https://github.com/pkos/RetroArch/pull/3 
.... Feel like this would need a bit more code review from some others, but it definitely was functional! It scanned a bunch of disc media correctly. Nice work!

**Comment by RobLoach** [comment]
@bkoropoff Would you mind giving a review, if possible? I've seen you in the scanning code too.

**Comment by i30817** [comment]
Aren't the segacd serials kind of unreliable? I feel like they're probably in JIS enconding or something from how crazy is looking a what appears in redump webpages and what appears if you look at the bytes in a stream of a good dump.

For that matter, is the scanner even using encodings?

**Comment by pkos** [comment]
The serials are pretty reliable.  11 chars (no JIS) at offset 0x0193 or 0x0194.

Bad dumps will be scrubbed by the scanner although that is not logged.

File: tasks/task_database_cue.c [hunk 7]
```diff
@@ -54,16 +55,192 @@ struct magic_entry
...
+ */
+static int find_disc_number(const char* str1, int index)
+{
+   char disc;
+   int disc_number = 0;
+
+   disc = str1[index + 6];
+   disc_number = disc - '0';
+
+   if (disc_number >= 1)
```
Comment (Jamiras) [inline tasks/task_database_cue.c]: What is disks are alphanumeric? "(Disk A)" should this also check `disc_number <= 9`?

**Comment by pkos** [comment]
Hacks with the same serial as the original will be scanned as original.  Hackers should take note.

File: tasks/task_database_cue.c (hunk 7 above)
Comment (pkos) [reply tasks/task_database_cue.c]: Thanks for this review.  I will make these code changes as they are well written.  Should get around to it tonight.

File: tasks/task_database_cue.c [hunk 8]
```diff
@@ -54,16 +55,192 @@ struct magic_entry
...
+ * Replaces multiple spaces with a single space in a string.
+ */
+static void replace_multi_space_with_single_space(char *str)
+{
+   char *dest = str;
+
+   while (*str != '\0')
+   {
+      while (*str == ' ' && *(str + 1) == ' ')
+      str++;
```
[the remark this answers is not in the document]
Comment (RobLoach) [reply tasks/task_database_cue.c]:
``` suggestion
         str++;
```

**Comment by RobLoach** [comment]
@Jamiras Thanks for the review!

File: tasks/task_database_cue.c [hunk 9]
```diff
@@ -54,16 +55,192 @@ struct magic_entry
...
+
+      *dest++ = *str++;
+   }
+
+   *dest = '\0';
+}
+
+/**
+ * Removes leading and trailing spaces and tabs from the given string.
+ */
```
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c]: Going to keep this function as intended.

File: tasks/task_database_cue.c [hunk 10]
```diff
@@ -124,196 +301,650 @@ static int64_t get_token(intfstream_t *fd, char *token, uint64_t max_len)
...
-         if (mode_test != MODETEST_VAL)
-            is_mode1 = 1;
+   for (pos = 0; pos < 100000; pos++)
+   {
+      intfstream_seek(fd, pos, SEEK_SET);
+
+      if (intfstream_read(fd, game_id, 5) > 0)
+      {
+         game_id[5] = '\0';
+         if (
```
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c]: Going to keep this a 4 byte string comparison as it will exactly match, where U or N would could be false positives.

File: tasks/task_database_cue.c (hunk 7 above)
Comment (pkos) [reply tasks/task_database_cue.c]: I put in a switch/case for Aa-Ii

File: tasks/task_database_cue.c [hunk 11]
```diff
@@ -124,196 +301,650 @@ static int64_t get_token(intfstream_t *fd, char *token, uint64_t max_len)
...
-         boot_file = tmp + 1;
-
-      tmp++;
+      if (intfstream_read(fd, region_id, 1) > 0)
+         region_id[1] = '\0';
+      else
+         return false;
+   }
+   else
+      return false;
```
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c]: Going to leave these, as format is consistent throughout, and this is where I could make a mistake.

File: tasks/task_database_cue.c [hunk 12]
```diff
@@ -124,196 +301,650 @@ static int64_t get_token(intfstream_t *fd, char *token, uint64_t max_len)
...
+   char pre_game_id[50];
+   char hyphen = '-';
 
-   if (!sub_channel_mixed)
+   for (pos = 0; pos < 100000; pos++)
    {
-      if (!(intfstream_tell(fp) & 0x7FF))
+      intfstream_seek(fd, pos, SEEK_SET);
+
+      if (intfstream_read(fd, raw_game_id, 12) > 0)
```
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c]: I'm unsure if i should just define a char[100000].. Suggestions or help?

File: tasks/task_database_cue.c [hunk 13]
```diff
@@ -54,16 +55,192 @@ struct magic_entry
...
+   int ctr = 0;
+   int index = -1;
+   int i;
+
+   for (i = 0; str[i] != '\0'; ++i) {
+      if (t == str[i])
+         ++ctr;
+   }
+
+   index = ctr;
```
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c:87]: done

File: tasks/task_database_cue.c [hunk 14]
```diff
@@ -54,16 +55,192 @@ struct magic_entry
...
+   while (str[ctr])
+   {
+      new_char = str[ctr];
+      if (isspace(new_char))
+         new_char = t;
+      *dest++ = new_char;
+      ctr++;
+   }
+
+   *dest = '\0';
```
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c:109]: done

File: tasks/task_database_cue.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c:122]: fixed

File: tasks/task_database_cue.c [hunk 15]
```diff
@@ -54,16 +55,192 @@ struct magic_entry
...
+   }
+
+   index = ctr;
+   return index;
+}
+
+/**
+ * Replaces all spaces with the given character.
+ */
+static void replace_space_with_single_character(char *str, char t)
```
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c:94]: done

File: tasks/task_database_cue.c [hunk 16]
```diff
@@ -54,16 +55,192 @@ struct magic_entry
...
+   for(j=0;s[i];i++)
+   {
+      s[j++]=s[i];
+   }
+   s[j]='\0';
+   for(i=0;s[i]!='\0';i++)
+   {
+      if(s[i]!=' '&& s[i]!='\t')
+         j=i;
+   }
```
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c:148]: done

File: tasks/task_database_cue.c [hunk 17]
```diff
@@ -54,16 +55,192 @@ struct magic_entry
...
+
+/**
+ * Find the position of a substring in a string.
+ */
+static int find_index_substring_string(const char* str1, const char* str2)
+{
+   int index;
+
+   int dlenstr = strlen(str1);
+   if (dlenstr > 0)
```
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c:189]: removed the length check and replaced as above... im not sure which functions handle empty strings so I prefer to keep the if statement.

File: tasks/task_database_cue.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c:315]: Trying the pointer method after reading in 100000 bytes.. testing

File: tasks/task_database_cue.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (Jamiras) [reply tasks/task_database_cue.c:380]: I wasn't suggesting removing the string comparisons. Since all of the string comparisons start with a U or N, you can avoid doing them if the next character is not a U or N.

File: tasks/task_database_cue.c [hunk 18]
```diff
@@ -124,196 +301,650 @@ static int64_t get_token(intfstream_t *fd, char *token, uint64_t max_len)
...
+   unsigned pos;
 
-         intfstream_read(fp, &mode_test, 4);
-         if (mode_test != MODETEST_VAL)
-            is_mode1 = 1;
+   for (pos = 0; pos < 100000; pos++)
+   {
+      intfstream_seek(fd, pos, SEEK_SET);
+
+      if (intfstream_read(fd, game_id, 5) > 0)
```
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c]: pointers are working

**Comment by i30817** [comment]
I have a question about this, does this update touch the thumbnail behavior? The usecase i'm interested in is the thumbnails when there is no serial or crc available.

I *think* it currently uses the filename, which is the reason the manual scanner even works for thumbnails sometimes as long as the actual name is the same as the database name.

Well i made a m3u script to create them : https://gist.github.com/i30817/ba37fbb2b3c6e34ff926ad833f465055

And my script makes multidisc games look like this first and not the second

`Xenogears (USA) (Disc 1).m3u`

`Xenogears (USA) (Disc 1).m3u`

this obviously misses the game thumbnail, though it's nicer. So i'm asking if there is any attempt to parse the serial of the first disc or something like that in m3u files if the name can't be used? Just to know if i'm going to have to have a compatibility switch on the tool to use that second form on m3u filenames, even if it's ugly.

**Comment by pkos** [comment]
> I have a question about this, does this update touch the thumbnail behavior? The usecase i'm interested in is the thumbnails when there is no serial or crc available.

When serials are available they will be matched to the database entry, this will effect the thumbnails as they are matched to the database entries.  No serial will generate a crc, if the crc is in the database it will effect thumbnails.  No crc, well I don't know if a case where no crc is generated.

**Comment by Sanaki** [comment]
Thumbnails have nothing to do with this, nor should they. If a thumbnail exists for a standard m3u name, it's used. Some already are in the thumbnail repo, such as `Chrono Cross (USA)`. The correct solution would be to add more thumbnails to match standardized names like that.

**Comment by i30817** [comment]
> No crc, well I don't know if a case where no crc is generated.

The manual scanner - though i think @Sanaka is correct, this would be cleaner fixed in the database, even if only as a symlink to the cover of 'disc 1' or something.

**Comment by pkos** [comment]
> The serials are pretty reliable. 11 chars (no JIS) at offset 0x0193 or 0x0194.
> 
> Bad dumps will be scrubbed by the scanner although that is not logged.

Scrubbed discs are now logged.

**Comment by pkos** [comment]
> > No crc, well I don't know if a case where no crc is generated.
> 
> The manual scanner - it matches the label to a rom-name on the database for the playlist system, and in that case the label is the filename minus extension - though i think @sanaka is correct, this would be cleaner fixed in the database, even if only as a symlink to the cover of 'disc 1' or something. The names without the 'disc' could be autogenerated from the redump cues collection, so maybe it's a good idea to add a issue on libretro-database.

I agree, Sega/Mega CD can be cleaned up a little in the database.. the Sega/Mega CD dat file has a lot of duplicate serials for games from different regions.  Rob and I should be able to handle different Mega/Sega CD regions in the code and database.

These changes will not affect the manual scanner, only "Scan Directory" and "Scan File" as with manual a database check is not made.

File: tasks/task_database_cue.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c]: I see, first comparison checks U or N... if true then do the 4-5 byte comparison.  I'll put that in.

File: tasks/task_database_cue.c [hunk 19]
```diff
@@ -124,196 +301,650 @@ static int64_t get_token(intfstream_t *fd, char *token, uint64_t max_len)
...
    {
-      if (!(intfstream_tell(fp) & 0x7FF))
+      intfstream_seek(fd, pos, SEEK_SET);
+
+      if (intfstream_read(fd, raw_game_id, 12) > 0)
       {
-         unsigned int mode_test = 0;
+         raw_game_id[12] = '\0';
+         strncpy(pre_game_id, raw_game_id, 4);
+         pre_game_id[4] = '\0';
```
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c]: Didn't realize the function string_is_equal_fast was available but i've made use of it now, much cleaner.

File: tasks/task_database_cue.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c]: Done

File: tasks/task_database_cue.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c:589]: Changed to the inverted functions with consistency throughout each detect serial function.

File: tasks/task_database_cue.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c]: Done

File: tasks/task_database_cue.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (pkos) [reply tasks/task_database_cue.c:122]: Rather, I will use the already made function for all whitespace.

**Comment by RobLoach** [comment]
@pkos Looks like a git conflict. Should be an easy fix through the GitHub UI.

**Comment by twinaphex** [comment]
Hmm, seems there is a merge conflict that has to be resolved first.

After the merge conflict is resolved, I can ask jdgleaver to review this.

**Comment by RobLoach** [comment]
Might have missed a few things in the conflicts... Would appreciate a review. https://github.com/pkos/RetroArch/pull/5

**Comment by twinaphex** [comment]
Hi guys, I am still very interested in this PR, but it hasn't seem to have gotten any traction in the past couple of months. Any initiative to get the remaining issues sorted?

**Comment by pkos** [comment]
> Hi guys, I am still very interested in this PR, but it hasn't seem to have gotten any traction in the past couple of months. Any initiative to get the remaining issues sorted?

Currently, the serial scanner is complete for the above systems, due to limitations in some digital media not all games will have a serial.  I'd say the serial scanner is a valid way to quickly scan games, however, it can be further improved in two ways.

First, the addition of "fuzzy" matching using the Levenshtein Distance could improve identification of games with the same serial but are actually considered different by Redump (ie. RE1, Beta, etc.).  In addition, after updating the RA code, there would need to be a simultaneous update to the RA databases to include these different games.

Second, a vast and complete identification of derivatives of the original Redump images is underway and slowly growing with support currently for SNK Neo Geo CD, NEC - PC - FX, and Panasonic 3DO.  The datafiles for these CRC32, internal SHA1, and MD5 are available here (https://github.com/pkos/CRoSG).  With these files added to the RA databases, it is possible to scan by crc32 the CHDs, and with updated RA code it would be improved to internal SHA1, the fastest and more versatile way to identify CHDs.

My final thought is that the serial scanner alone should be working as complete as it is possible to make it and it will suffer some limitations to the number of game it will identify alone.

**Comment by pkos** [comment]
Also, I've forked RA with the latest code updates here (https://github.com/pkos/RetroArch).

**Comment by i30817** [comment]
Levenshtein Distance is not a good idea to id gamecube games. Changing a letter in the first position is much more meaningful that changing many letters in the last.

Honestly it should be hardcoded, and if you require a 'group' for the variants for reasons and it's not predictable from the Nintendo standard, just group them on a fileformat somewhere.

I'd also mention that 'games hacks being id'ed as the original game' is a very common problem on other platforms that use this kind of flawed id, but it's admittedly not much of a problem on gamecube/wii because there are few hacks (some translations but it doesn't affect the settings needing to be different most times).

**Comment by pkos** [comment]
> Levenshtein Distance is not a good idea to id gamecube games. Changing a letter in the first position is much more meaningful that changing many letters in the last and levenshtein is geared to pick up mistakes which are independent of position or to identify languages (with a weird 'average levensthtein distance' thing that works most of the time surprisingly).
> 
> Honestly it should be hardcoded, and if you require a 'group' for the variants for reasons and it's not predictable from the Nintendo standard, just group them on a fileformat somewhere.
> 
> I'd also mention that 'games hacks being id'ed as the original game' is a very common problem on other platforms that use this kind of flawed id, but it's admittedly not much of a problem on gamecube/wii because there are few hacks (some translations but it doesn't affect the settings needing to be different most times).
> 
> edit: rereading the post it appears you don't want to use the algorithm on the serial? Why would you want fuzziness then?

I think you've confused the scope of the LD with actually matching by name, this is not the case, it will only be done by serial or CRC32.  Same serials however would benefit from LD (limited scope).

**Comment by RobLoach** [comment]
The latest changes are over in this Pull Request: https://github.com/libretro/RetroArch/pull/11719
